### DISCUSSION TITLE
Using PageResolver to pass data in Shell.

### DISCUSSION THREAD

**Discussion by spuliy** [matt-goldman/Maui.Plugins.PageResolver#37] [Q&A] [answered]
Hello, I was wondering if I could use this plugin to pass data instead of using query parameters in Shell navigation. For example:

`Shell.Current.Navigation.PushAsync<MyPage>(new Item { Property = "Something" });`

Navigating to Page:
```

      public MyPage(MyPageModel viewmodel, Item value)
	{
            InitializeComponent();
            
            BindingContext = viewmodel;
  
            
            viewmodel.Init(Item);
        }
    
```

I get an exception when I try this: 

`'A suitable constructor for type 'App.Pages.MyPage' could not be located. Ensure the type is concrete and all parameters of a public constructor are either registered as services or passed as arguments. Also ensure no extraneous arguments are provided.'`

But I'm not sure if that is because I'm not allowed to do this if I am using Shell or I am doing something wrong. I saw a question similar to this but I was not sure if it meant this was allowed in Shell Navigation or not.

**Comment by IeuanWalker** [comment]
Have you registered MyPage and your ViewModel with the DI container?

**Reply by spuliy** [reply]
I believe so. This is a summarized format of what I have registered:
```
	public static MauiApp CreateMauiApp()
	{
	  var builder = MauiApp.CreateBuilder();
	  builder
		  .UseMauiApp<App>()
		  .RegisterPageModels()
		  .RegisterPages()
		  .UsePageResolver();

	  MauiApp mauiapp = builder.Build();

	  return mauiapp;
	}

	public static MauiAppBuilder RegisterPageModels(this MauiAppBuilder mauiAppBuilder)
	{
	  mauiAppBuilder.Services.AddTransient<MyPageModel>();
	  return mauiAppBuilder;
	}

	public static MauiAppBuilder RegisterPages(this MauiAppBuilder mauiAppBuilder)
	{
	  mauiAppBuilder.Services.AddTransient<MyPage>();

	  return mauiAppBuilder;
	}
```

**Reply by IeuanWalker** [reply]
@spuliy personally im not sure you need to use this package if your using shell, as it has dependency resolution built in _(i dont use shell though, so cant really comment on this)_ - https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/dependency-injection?view=net-maui-8.0#automatic-dependency-resolution

Does it work if you just use -
```csharp
Navigation.PushAsync<MyPage>(new Item 
{ 
    Property = "Something"
 });
```

Instead of -
```csharp
Shell.Current.Navigation.PushAsync<MyPage>(new Item 
{ 
    Property = "Something"
 });
```

**Reply by matt-goldman** [reply, marked as answer]
> @spuliy personally im not sure you need to use this package if your using shell, as it has dependency resolution built in _(i dont use shell though, so cant really comment on this)_ - https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/dependency-injection?view=net-maui-8.0#automatic-dependency-resolution
> 
> Does it work if you just use -
> 
> ```cs
> Navigation.PushAsync<MyPage>(new Item 
> { 
>     Property = "Something"
>  });
> ```
> 
> Instead of -
> 
> ```cs
> Shell.Current.Navigation.PushAsync<MyPage>(new Item 
> { 
>     Property = "Something"
>  });
> ```

Agree here - using `Shell.Current.Navigation.PushAsync` doesn't really make sense. `Shell` is within the navigation stack (where you would use `PushAsync`), whereas navigation _within_ Shell would typically use `GoToAsync` which, as @IeuanWalker points out, has dependency resolution already built in.

Additionally, if you're navigating with Shell, you would also typically use query parameters or navigation state to pass values. https://learn.microsoft.com/en-us/dotnet/maui/fundamentals/shell/navigation?view=net-maui-8.0#pass-data

Again as @IeuanWalker says I'm not sure this package is right for your use case, but if you can elaborate a bit more on what you're doing here happy to help out. One use case here might be if you're trying to push a modal page, that might make sense. Otherwise combining hierarchical navigation with Shell is probably just making problems for yourself.

**Comment by spuliy** [comment]
Thank you @IeuanWalker and @matt-goldman for the advice, you guys are correct. Shell did have mostly everything I needed. 

The only problem I had was not being able to resolve a popup page using Mopups, but using the Mopups extension in this package I was able to resolve that issue. :)
